ecosystem/sep-0010.md

```diff
@@ -149,13 +164,17 @@ To validate the challenge transaction the following steps are performed by the s
   * verify that client signature is correct for the master key of the account address;
 * verify that transaction sequenceNumber is equal to zero;
 * use operations's source account to determine the authenticating client and perform any additional service-specific validations.
+* if the transaction has a memo:
+  * verify the transaction's memo and source account can be used to internally identify the authenticated user;
 
 The verification process confirms that a user holds an account. Depending on your application this may mean complete signing authority, some threshold of control, or being a signer of the account. See [Verification](#verification) for examples.
 
 Upon successful verification, service responds with a session JWT, containing the following claims:
 
 * `iss` (the principal that issued a token, [RFC7519, Section 4.1.1](https://tools.ietf.org/html/rfc7519#section-4.1.1)) — a [Uniform Resource Identifier (URI)] for the issuer (`https://example.com` or `https://example.com/G...`)
-* `sub` (the principal that is the subject of the JWT, [RFC7519, Section 4.1.2](https://tools.ietf.org/html/rfc7519#section-4.1.2)) — the public key of the authenticating Stellar account (`G...`)
+* `sub` (the principal that is the subject of the JWT, [RFC7519, Section 4.1.2](https://tools.ietf.org/html/rfc7519#section-4.1.2)) — the public key of the authenticating Stellar account potentially contcatenated with the `memo` value originally specified by the client using an `-` character
+  * For example, when `account` is pooled, the user can be identified using the 32-byte hash memo: `GAMJ4W6QDEKJWZRMAWGD7ETKPIMKUX4NCZ3JBEYPILZSOWTW7NL6TDQ4-yQ2eD5xHRSWUFJfCZqggCS5peHFkb9SXJ3hBjjZjHnc=`
+  * Web services that are passed a JWT token can use the `sub` memo value to internally identify the user who owns the funds held by the client on each request
```

Comment:
Unfortunately the `sub` field must be a string limiting us to the delimiter approach, but we do not need to use `sub`. This could be a good time to make this a breaking change where by we abandon the use of the `sub` field and define new fields like `account` and `memo`. Something like:

Standard account:
```
"account": "G..."
```

Pooled or multiplexed account:
```
"account": "M..."
```
or
```
"account": {
    "account": "G...",
    "memo": "1234"
}
```

Reply:
I'm usually not one to back down from a breaking change but SEP-10 is one of the most widely used standards in the Stellar ecosystem, and given the current state of SEP-10 regarding the 2.0 SDK updates, I think we should try not to break things until we define a better transition process.

Reply:
Yeah that's a good point.

I worry that delimiting the data in `sub` will come back to bite us in ways we don't expect, but I don't have any concrete examples.

I like that with this proposal the SDKs we know of should break if they see the hyphen and haven't been upgraded to support a memo. So they shouldn't accidentally validate a JWT as proving control of an account when it only proves alignment with a memo value.

Some other options:
- We could also define the `sub` to be encoded in some other format, such as JSON, so it is JSON stringified. This would make it predictable using common technology to build and parse. The value of sub would become:
   ```
   "sub": "{\"acc\":\"G...\",\"memo\":\"1234\"}"
   ```
   This makes it clear for any developer looking at the data at what they have because it is self-describing.
- We could use the strkey multiplexed format in SEP-10 without supporting them generally in other ways. This would allow us to adopt SEP-10 in the direction of that standard without waiting for any other part of it being implemented. We'd only use it to encode the data we have. This would also limit the memo types to the 64-bit ID.

Reply:
I'm not familiar with the strkey multiplexed format I'll take a look at it and come back with comments/questions.

Regarding the first 'other' option, it would certainly work, but I think it may confuse clients. If they've already deserialized the JSON string from the response body, it seems strange to deserialize a specific attribute within the JWT.

Personally I think the current design is the way to go. The one concern I have is that the current design doesn't provide the `memo_type` value in the `sub` field of the JWT. I think we need a better format for encoding this info.

The SDK's would be updated to validate this format if present, otherwise the original Stellar public key value would be expected.

Reply:
We can avoid needing to include the memo type if we limit the memo types allowed to the memo types allowed by multiplexed addresses / muxed accounts, which is the memo id type, a 64-bit integer. If we're thinking ahead to when multiplexed addresses are in use we'll be limited to those memo types anyway. My understanding is the research @stanford-scs did highlighted that most folks only needed that memo type. @tomerweller Thoughts?

Reply:
@tomerweller bump, starting to think about this again

Reply:
Yes. We discussed this during last year's meridian protocol session and there was a wide agreement that int64s are enough. Which will make transitioning to muxed accounts easier.